hal_nordic: nrfs: dvfs: Update SystemCoreClock on frequency change

nordic-krch opened the pull request:
Update SystemCoreClock to match current core frequency. This MDK variable is used by nrfx_coredep_delay_us which performs busy waiting and is used by k_busy_wait.

Without that when DVFS is enabled then `k_busy_wait` is inaccurate, up to 5 times slower than expected when the lowest operation point is used (64MHz).

hubertmis reviewed (commented):
I believe `SystemCoreClock` should be modified only when `ld_dvfs_configure_hsfll()` succeeds. Also, it seems this is not the only place in which HSFLL gets reconfigured. Another one should be before starting the downscaling procedure. So setting `SystemCoreClock` should be there as well.

Or maybe it should be inside `ld_dvfs_configure_hsfll`?

nordic-krch committed:
hal_nordic: nrfs: dvfs: Update SystemCoreClock on frequency change

Update SystemCoreClock to match current core frequency. This MDK
variable is used by nrfx_coredep_delay_us which performs busy
waiting and is used by k_busy_wait.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

lstnl commented:
> @lstnl @Rafal-Nordic can you point us to the best place where new frequency is settled?

There is callback function dvfs_service_handler_register_freq_setting_applied_callback which is used in clock control to indicate when new frequency setting was applied. When this is called new frequency should be there.

nordic-krch commented:
@lstnl can you take a look now? I've added setting this variable to 2 places from which this callback can be called.

lstnl commented on modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c:
```diff
@@ -139,9 +139,19 @@ static void dvfs_service_handler_scaling_background_job(enum dvfs_frequency_sett
 }
 
+/* Update MDK variable which is used by nrfx_coredep_delay_us (k_busy_wait). */
+static void dvfs_service_update_core_clock(enum dvfs_frequency_setting oppoint_freq)
```
This would need to be called every time ld_dvfs_configure_hsfll is called so also init and dvfs_service_handler_prepare_to_scale. I would move this inside ld_dvfs_configure_hsfll function, what do You think?
